File: Java/src/napakalaki/Napakalaki.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;
import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author ludvins
 */
public class Napakalaki {
    
    private static Napakalaki instance = null;
    private Monster currentMonster;
    private final CardDealer cardDealer = CardDealer.getInstance();
    private int currentPlayer;
    private ArrayList<Player> players; 
    
    private Napakalaki () {} 
    
    public static Napakalaki getInstance () {
        
        if (instance == null) instance = new Napakalaki();
       
        return instance;
    }
    
    private void initPlayers(ArrayList<String> names){
        
        this.players = new ArrayList();

        for (String a : names){
            this.players.add(new Player(a));
        }
        
        Random rd = new Random();
        this.currentPlayer = rd.nextInt(this.players.size());
        
    }
    
    private void nextPlayer(){
        
       this.currentPlayer = (this.currentPlayer + 1) % this.players.size();
    }
    
    private boolean nextTurnIsAllowed(){
 
        return this.players.get(currentPlayer).ValidState();

    }
    
    private void setEnemies() {
        
        Random rd = new Random();
        int n;
        
        for (int i = 0; i < this.players.size(); i++){
     
            do {
            
                n = rd.nextInt( this.players.size());
        
            }
            while(n == i); 
            
            players.get(i).setEnemy(this.players.get(n));
        }
    }
    
    public CombatResult developCombat(){
        
        CombatResult combat = this.players.get(currentPlayer).combat(this.currentMonster);
        
        if (combat == CombatResult.LOSEANDCONVERT){
            CultistPlayer c = new CultistPlayer(this.players.get(this.currentPlayer), this.cardDealer.nextCultist());
            
            for (Player p : this.players){
                
                if (p.getEnemy() == players.get(currentPlayer)) p.setEnemy(c);
                
            }
            
            players.set(currentPlayer, c);
            
        }
        
        return combat;
    }
    
    public void discardVisibleTreasures(ArrayList<Treasure> treasures){
        
        for (Treasure t : treasures){
            
            this.players.get(currentPlayer).discardVisibleTreasure(t);
            
        }
        
    }
    
    public void discardHiddenTreasures(ArrayList<Treasure> treasures){
        
        for (Treasure t: treasures){
            
            this.players.get(currentPlayer).discardHiddenTreasure(t);
        
        }
    }
    
    public void makeTreasureVisible(ArrayList<Treasure> treasure){
        
        for (Treasure t : treasure){
            
            this.players.get(currentPlayer).makeTreasureVisible(t);
        }
    }
    
    public void initGame (ArrayList<String> names) {
        
        this.initPlayers(names);
        
        this.setEnemies();
        
        this.cardDealer.initCards();
        
        this.nextTurn();
    }
    
    public Player getCurrentPlayer(){
        
        return this.players.get(currentPlayer);
    }
    
    public Monster getCurrentMonster() {
        
        return this.currentMonster;
        
    }
    
    public boolean nextTurn(){
        
        boolean ret = this.nextTurnIsAllowed();
        
        if(ret){
            
            this.currentMonster = this.cardDealer.nextMonster();
            this.nextPlayer();
                     
            if (this.players.get(currentPlayer).isDead())this.players.get(currentPlayer).initTreasures();
            
        }
        
        return ret;
    }
    
    public boolean endOfGame(CombatResult cr){
        
        return cr == CombatResult.WINGAME;
    }
    
}

File: Ruby/napakalaki/lib/napakalaki.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module NapakalakiGame

require_relative 'bad_consequence.rb'
require_relative 'prize.rb'
require_relative 'monster.rb'
require_relative 'treasure.rb'
require_relative 'player.rb'
require_relative 'card_dealer.rb'
require_relative 'cultist.rb'
require_relative 'cultist_player.rb'
require "singleton"

class Napakalaki
  
  include Singleton
  
  private
  
  def initPlayers(names)
    @players = Array.new
    
    names.each do |a|
      @players << Player.new(a)
    end
  end
  
  def nextPlayer
    
    if @currentPlayer == nil
      @currentPlayer = rand(@players.size)
    else 
      
      @currentPlayer = (@currentPlayer + 1)% @players.size
      
    end
    
    
    
  end
  
  def nextTurnAllowed
    
    if @currentPlayer.nil?
      return true
    end
    
    return @players[@currentPlayer].validState

  end
  
  def setEnemies
    
    r = Random.new
    
    for i in 0..@players.size - 1
    
      loop do
            
        r = rand(@players.size)
    
   
        break if @players[i] != @players[r]
    
      end
    
      @players[i].setEnemy(@players[r])
    
    end
  end
  
  public
  
  def developCombat
    
    result = @players[@currentPlayer].combat(@currentMonster)
    
    if result == CombatResult::LOSEANDCONVERT
      
      cultist = CultistPlayer.new(@players[@currentPlayer], @dealer.nextCultist)
      
      @players.each do |p|
        
        p.setEnemy(cultist) if p.getEnemy == @players[@currentPlayer]
        
      end
      
      @players[@currentPlayer] = cultist

    end
  
    return result
    
  end
  
  def discardVisibleTreasures(treasures)
    
    treasures.each do |t|
      @players[@currentPlayer].discardVisibleTreasure(t)
    end
    
  end
  
  def discardHiddenTreasures(treasures)
    
    treasures.each do |t|
      @players[@currentPlayer].discardHiddenTreasure(t)
    end
  end
  
  def makeTreasuresVisible(treasures)
    
    treasures.each do |t|
      
      @players[@currentPlayer].makeTreasureVisible(t)
    end
    
  end
  
  def initGame(names)
    
    @dealer = CardDealer.instance
    
    @dealer.initCards
    
    initPlayers(names)
    
    setEnemies
   
    nextTurn
    
  end
  
  def getCurrentPlayer
    @players[@currentPlayer]
  end
  
  def getCurrentMonster
    @currentMonster
  end
  
  def nextTurn
    
    ret = nextTurnAllowed

     if ret then
       @currentMonster = @dealer.nextMonster
       nextPlayer

      if @players[@currentPlayer].isDead then
        @players[@currentPlayer].initTreasures
      end
     end
    return ret
  end
  
  def endOfGame(cr)
    
    cr == CombatResult.WINGAME
    
  end
end
end
